### PR TITLE
Remember when a signed in user acknowledges the cookie law.

### PR DESCRIPTION
**Pull request by jfly**
This is nicer than asking them every time they sign into a new device or
reset their cookies.

In order to test this, I re-enabled the consent banner everywhere
(reverting @viroulep's suggestion here:
https://github.com/thewca/worldcubeassociation.org/pull/6327#discussion_r663331261).
I found that this only broke tests in features/sign_up_spec.rb, so it
seems fine to me to leave enabled everywhere. Maybe this went so
smoothly because I changed our user factory to default to
`cookies_acknowledged=true`? I didn't investigate too far.

Sorry for the unrelated seeming changes in `.eslintrc.json`. I wanted to
log some sort of warning when the user acknowledges the cookie law but
they aren't logged in yet (or are no longer logged in). I figured it
would be reasonable to log that using `console.warn`. I also opted to
move the no-console rule from `application.js` into `.eslintrc.json`,
which feels like a better place to me. Lastly, I changed the eslintrc
rules from using numbers (0, 1, 2) to explicitly using strings ("off",
"warn", "error"), which seems way clearer to me. Happy to revert any of
this or pull it out into a separate PR if folks prefer.

### CODE REVIEW

**Comment by jfly** [comment]
@gregorbg looks like the [tests failed in unrelated code](https://github.com/thewca/worldcubeassociation.org/pull/6384/checks?check_run_id=3175990622#step:12:3922):

```
  1) Strong Customer Authentification payment on the 'register' page of a visible competition user fails to complete the 3D secure challenge FIXED
     Expected pending 'figure out why iframes suddenly don't get rendered in headless mode any more' to fail. No error was raised.
     # ./spec/features/sca_payments_spec.rb:28

  2) Strong Customer Authentification payment on the 'register' page of a visible competition user pays with a 3D secure enabled card FIXED
     Expected pending 'figure out why iframes suddenly don't get rendered in headless mode any more' to fail. No error was raised.
     # ./spec/features/sca_payments_spec.rb:23
```

Is it possible whatever weirdness you identified in https://github.com/thewca/worldcubeassociation.org/pull/6380 is no longer an issue?

**Comment by gregorbg** [comment]
It is _possible_, but just now I tried reverting the commit where I marked the tests as pending and to my surprise the CI failed because the tests didn't proceed. This is starting to look like undeterministic failures which must have a different reason than what I initially suspected.

For reference, see my draft at https://github.com/thewca/worldcubeassociation.org/pull/6387

File: WcaOnRails/.eslintrc.json [hunk 1]
```diff
@@ -25,10 +25,11 @@
         "jquery"
     ],
     "rules": {
```
Comment (gregorbg) [inline WcaOnRails/.eslintrc.json:27]: What are these changes about? Do they have anything to do with your changes to the cookie banner?

File: WcaOnRails/Gemfile.lock [hunk 2]
```diff
@@ -468,7 +474,7 @@ GEM
       thor (>= 0.19.0, < 2.0)
     rainbow (3.0.0)
     raindrops (0.19.1)
-    rake (13.0.3)
+    rake (13.0.6)
```
Comment (gregorbg) [inline WcaOnRails/Gemfile.lock:477]: I'd rather not "sideload" dependency updates like this under the radar.

File: WcaOnRails/Gemfile.lock [hunk 3]
```diff
@@ -738,4 +744,4 @@ DEPENDENCIES
   wicked_pdf
 
 BUNDLED WITH
-   2.1.4
+   2.2.24
```
Comment (gregorbg) [inline WcaOnRails/Gemfile.lock:747]: Please either
- bundle with 2.1.4, so that this version number doesn't constantly change with every PR that members with different local `Bundler` versions end up creating
- or share your Ruby wisdom on how to avoid/suppress this line altogether, it has been causing us trouble with onboarding in the past (people had to deliberately install bundler 2.1.4 instead of "just" the latest bundler)

File: WcaOnRails/db/structure.sql [hunk 4]
```diff
@@ -1687,6 +1688,7 @@ INSERT INTO `schema_migrations` (version) VALUES
 ('20200125180554'),
 ('20200206012756'),
 ('20200304044931'),
+('20200319193625'),
```
Comment (gregorbg) [inline WcaOnRails/db/structure.sql:1691]: good catch!

File: WcaOnRails/spec/features/sign_up_spec.rb [hunk 5]
```diff
@@ -7,6 +7,13 @@
   let!(:person_without_dob) { FactoryBot.create :person, year: 0, month: 0, day: 0 }
   let!(:custom_delegate) { FactoryBot.create(:delegate) }
 
+  before :each do
+    # The cookie banner just gets in the way of these tests, and is already
+    # tested elsewhere. Set a cookie that prevents the cookie banner from
+    # appearing.
+    page.driver.set_cookie('cookie_eu_consented', 'true')
```
Comment (gregorbg) [inline WcaOnRails/spec/features/sign_up_spec.rb:14]: Cool! We had been struggling to hide the cookie bar in testing before (the brute-force solution was to only render the cookie bar if `Rails.production?` returned `true`), would you mind sharing how you found out about this trick?

**Review by gregorbg — CHANGES_REQUESTED** [review]
Thank you for your contribution, it's great to see you're still around :)
I have added a few remarks about your implementation, but I guess the only real "problem" for now is a `bundler` version conflict in the `Gemfile.lock`.

Also, I am generally a bit reluctant to forking popular gems only for the sake of prematurely merging one PR. I understand that this is very convenient and easy to undo, but as long as the original maintainer doesn't provide a release you're effectively committing to maintaining your fork. This can lead to additional work down the road if the relase takes longer than anticipated and you have to stay up-to-date with upstream changes. Are you willing to commit to this additional technical debt?

File: WcaOnRails/Gemfile.lock (hunk 2 above)
Comment (jfly) [reply WcaOnRails/Gemfile.lock:477]: Aha! Sorry, totally missed this. I ran `bundle update cookies_eu`, but apparently the right incantation is `bundle update --conservative cookies_eu` (from https://stackoverflow.com/a/23268530/1739415)

File: WcaOnRails/.eslintrc.json (hunk 1 above)
Comment (jfly) [reply WcaOnRails/.eslintrc.json:27]: Sort of. It's part related to the `console.warn` I added, and part unrelated cleanup. Please see the last paragraph of my commit message ("Sorry for the unrelated seeming changes in .eslintrc.json...") for a full explanation.

File: WcaOnRails/Gemfile.lock (hunk 3 above)
Comment (jfly) [reply WcaOnRails/Gemfile.lock:747]: Ah good point. I've installed bundler 2.1.4 to get rid of this part of the diff.

I unfortunately don't know what the "ruby way" is of handling this sort of thing. It is super annoying how easy it is for it to change when you have a bunch of people all working on the same repo. Is bundler supposed to become a core part of ruby someday...?

File: WcaOnRails/spec/features/sign_up_spec.rb (hunk 5 above)
Comment (jfly) [reply WcaOnRails/spec/features/sign_up_spec.rb:14]: That was tricky! My process went something like:

1. Google "rspec feature spec cookies". Give up after reading [this](https://stackoverflow.com/questions/29531499/how-to-test-cookies-with-rspec-and-capybara-in-a-feature-test) and [this](https://relishapp.com/rspec/rspec-rails/docs/controller-specs/cookies) (this is for controller specs, not feature specs, thanks google!)
2. Stop and think a bit. We use phantomjs, right?
3. Oh no [we don't anymore](https://github.com/thewca/worldcubeassociation.org/commit/b953fbb2e6f36ca4df0e22b9eec61df38528dfa8). Thanks @gregorbg!
4. Google "capybara apparition cookies".
5. Find https://github.com/twalpole/apparition#manipulating-cookies. Oooh this looks promising!

**Review by jfly — COMMENTED** [review]
> Also, I am generally a bit reluctant to forking popular gems only for the sake of prematurely merging one PR.

I totally agree. Fortunately it's a non-issue, as the cookies_eu folks have already merged my PR and released a new version of their gem! I've updated the diff to not rely upon my fork anymore.

File: WcaOnRails/.eslintrc.json (hunk 1 above)
Comment (gregorbg) [reply WcaOnRails/.eslintrc.json:27]: Whoops, my bad for missing that part of your PR description.

File: WcaOnRails/spec/features/sign_up_spec.rb (hunk 5 above)
Comment (gregorbg) [reply WcaOnRails/spec/features/sign_up_spec.rb:14]: Thanks, good to know! I also just realised that the name of the cookie is buried in [the gem's README](https://github.com/infinum/cookies_eu#cookies-used)

File: WcaOnRails/.eslintrc.json (hunk 1 above)
Comment (gregorbg) [reply WcaOnRails/.eslintrc.json:27]: Seeing that the previous configuration was set in `application.js` as
```js
/* eslint no-console:0 */
```
I wonder how you came up with this new
```js
"no-console": ["error", { "allow": ["warn", "error"] }]
```
config value. Can you please explain what the `allow` block even does and why it contains `error`? To me this reads a bit like "no console for errors but allow errors"?!

File: WcaOnRails/app/assets/javascripts/application.js [hunk 6]
```diff
@@ -366,4 +366,3 @@ $(function() {
     form.submit();
   });
 });
-cookiesEu.init()
```
Comment (gregorbg) [inline WcaOnRails/app/assets/javascripts/application.js:369]: Is this change intentional?

**Review by gregorbg — CHANGES_REQUESTED** [review]
Thanks for your prompt fixes, and also sorry for missing the explanation of the `.eslintrc.json` changes in your commit message :see_no_evil:
Still have a few open questions but once these are resolved we are good to go!

File: WcaOnRails/app/assets/javascripts/application.js (hunk 6 above)
Comment (jfly) [reply WcaOnRails/app/assets/javascripts/application.js:369]: Oh yes. I don't know why we put this here in the first place, because [`cookiesEu` automatically invokes it's `init` function when the page loads](https://github.com/infinum/cookies_eu/blob/3a0ad36b51b4be0b038c41c89912ea301ecfb90a/app/assets/javascripts/cookies_eu.js#L52). In fact, this `init` method is totally undocumented, so I don't think there's any guarantee that they keep it around, as it's just an implementation detail.

File: WcaOnRails/.eslintrc.json (hunk 1 above)
Comment (gregorbg) [reply WcaOnRails/.eslintrc.json:27]: It also seems reasonable to me, thanks for the thorough explanation!
Also I totally share your opinion about string values being more concise and better readable than their `0|1|2` int counterparts

**Review by gregorbg — APPROVED** [review]
LGTM :ship:
